src/sources/dnstap/dns_message_parser.rs

```diff
@@ -0,0 +1,2000 @@
+    }
+}
+
+fn format_bytes_as_hex_string(bytes: &[u8]) -> String {
+    bytes
+        .iter()
+        .map(|e| format!("{:02X}", e))
+        .collect::<Vec<String>>()
+        .join(".")
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{
+        net::{Ipv4Addr, Ipv6Addr},
+        str::FromStr,
+        time::Instant,
+    };
+    use trust_dns_proto::{
+        rr::{
+            dnssec::{
+                rdata::{
+                    dnskey::DNSKEY, ds::DS, nsec::NSEC, nsec3::NSEC3, nsec3param::NSEC3PARAM,
+                    sig::SIG, DNSSECRData,
+                },
+                Algorithm as DNSSEC_Algorithm, DigestType, Nsec3HashAlgorithm,
+            },
+            domain::Name,
+            rdata::{
+                caa::KeyValue,
+                sshfp::{Algorithm, FingerprintType},
+                tlsa::{CertUsage, Matching, Selector},
+            },
+            rdata::{null, CAA, NAPTR, SSHFP, TLSA, TXT},
+        },
+        serialize::binary::Restrict,
+    };
+
+    impl DnsMessageParser {
+        pub fn raw_message_clone(&self) -> Option<&Vec<u8>> {
+            self.raw_message_clone.as_ref()
+        }
+    }
+
+    #[test]
+    fn test_parse_as_query_message() {
+        let raw_dns_message = "szgAAAABAAAAAAAAAmg1B2V4YW1wbGUDY29tAAAGAAE=";
+        if let Ok(raw_qeury_message) = base64::decode(raw_dns_message) {
+            let parse_result = DnsMessageParser::new(raw_qeury_message).parse_as_query_message();
+            assert!(parse_result.is_ok());
+            if let Ok(message) = parse_result {
+                assert_eq!(message.header.qr, 0);
+                assert_eq!(message.question_section.len(), 1);
+                assert_eq!(
+                    message.question_section.first().unwrap().name,
+                    "h5.example.com."
+                );
+                assert_eq!(
+                    &message
+                        .question_section
+                        .first()
+                        .unwrap()
+                        .record_type
+                        .clone()
+                        .unwrap(),
+                    "SOA"
+                );
+            } else {
+                error!("Message is not parsed.");
+            }
+        } else {
+            error!("Invalid base64 encoded data.");
+        }
+    }
+
+    #[test]
+    fn test_parse_as_query_message_with_invalid_data() {
+        if let Err(err) = DnsMessageParser::new(vec![1, 2, 3]).parse_as_query_message() {
+            assert!(err.to_string().contains("unexpected end of input"));
+            match err {
+                DnsMessageParserError::TrustDnsError { source: e } => {
+                    assert!(e.to_string().contains("unexpected end of input"))
+                }
+                DnsMessageParserError::SimpleError { cause: e } => {
+                    error!("Expected TrustDnsError, got {}.", &e)
+                }
+                _ => error!("{}.", err),
```

Comment:
This is just using the tracing macro, right? I think you want to actually fail the test if these `error!` lines are hit. You can just use `panic`. Same for below tests.

Reply:
Yes will use `panic!` instead for all related tests.

Reply:
Updated